src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java

```diff
@@ -80,19 +92,119 @@ public void completed(Integer result, Void v) {
+            buffer.put(entry.data);
+            buffer.flip();
+            afc.write(
+                    /* source= */ buffer,
+                    /* position= */ 0,
+                    /* attachment= */ null,
+                    new CompletionHandler<Integer, Void>() {
+                        @Override
+                        public void completed(Integer resultLen, Void ignore) {
+                            if (closeChannel(afc, "completed read")) {
+                                if (resultLen != size) {
+                                    VolleyLog.e(
+                                            "File changed while reading: %s",
+                                            file.getAbsolutePath());
+                                    callback.onPutComplete();
+                                    return;
+                                }
+                                header.size = resultLen;
+                                mTotalSize =
+                                        DiskBasedCacheUtility.putEntry(
+                                                key, header, mTotalSize, mEntries);
+                                mTotalSize =
+                                        DiskBasedCacheUtility.pruneIfNeeded(
+                                                mTotalSize,
+                                                mMaxCacheSizeInBytes,
+                                                mEntries,
+                                                mRootDirectorySupplier);
+                            }
+
+                            callback.onPutComplete();
+                        }
+
+                        @Override
+                        public void failed(Throwable throwable, Void ignore) {
+                            VolleyLog.e(
+                                    throwable, "Failed to read file %s", file.getAbsolutePath());
+                            callback.onPutComplete();
+                            closeChannel(afc, "failed read");
+                        }
+                    });
+        } catch (IOException e) {
+            if (closeChannel(channel, "IOException")) {
+                boolean deleted = file.delete();
+                if (!deleted) {
+                    VolleyLog.d("Could not clean up file %s", file.getAbsolutePath());
+                }
+                initializeIfRootDirectoryDeleted();
+            }
+            callback.onPutComplete();
+        }
+    }
+
+    public void initialize() {
+        // TODO (sphill99): #181
+    }
+
+    /** Re-initialize the cache if the directory was deleted. */
+    private void initializeIfRootDirectoryDeleted() {
+        if (mRootDirectorySupplier.get().exists()) {
+            return;
+        }
+        VolleyLog.d("Re-initializing cache after external clearing.");
+        mEntries.clear();
+        mTotalSize = 0;
+        initialize();
+    }
+
+    private boolean closeChannel(AsynchronousFileChannel afc, String endOfMessage) {
```

Comment:
consider a quick Javadoc, since the meaning of the return value is perhaps non-obvious.

Reply:
added javadoc